### PR TITLE
Download project source from CLI

### LINKED ISSUE #69: Download source bundle from CLI
`$ modulus project download` maybe?

This can use the same endpoint in the API as the portal.

### PR DESCRIPTION
**Pull request by harlanj**
- Add librarian call to api to get project source
- Add feature to download latest deployed source code from API
- Add flags to specify zip name when downloading source code

Closes #69

### CODE REVIEW

**Comment by tzmanics** [comment]
There are a few formatting things that would be great to fix so all new code going in is as pristine as possible 😁 Let's tackle them before merging.

File: lib/controllers/project.js [hunk 1]
```diff
@@ -588,6 +588,48 @@ Project.prototype.scale = function (projectId, instances, await, callback) {
...
+ * @param {string} outLocation - the path to the directory where the tarball should be downloaded
+ * @param {function} callback - function invoked with the file stream. Responsible for the actual download.
+ */
+//-----------------------------------------------------------------------------
+Project.prototype.download = function(projectId, projectName, outLocation, callback) {
+  librarian.project.download(projectId, userConfig.data.apiKey, function (err, req, res) {
+    if (err) return callback(Errors.getMessage(err));
+
+
+    var date = new Date().toISOString().replace(/T/, ' ').replace(/\..+/, '');
```
Comment (jackboberg) [inline lib/controllers/project.js]: Slightly simpler format string that we already use in `emmet`:

`new Date().toISOString().replace(/:/g, '.') // 2016-01-15T20.39.14.868Z`

**Comment by tzmanics** [comment]
:+1:

File: lib/controllers/project.js [hunk 2]
```diff
@@ -588,6 +588,50 @@ Project.prototype.scale = function (projectId, instances, await, callback) {
...
+    if (typeof outLocation === 'string') {
+      outFile = outLocation;
+      outFile = outFile.replace(/\//g, '/');
+      if (outFile.substring(outFile.length-4) !== '.zip') outFile += '.zip';
+
+      message = 'Directory ' + path.dirname(outLocation) + ' does not exist.';
+      if (!fs.existsSync(path.dirname(outLocation))) return callback(message);
+    }
+
+    dir = path.dirname(outLocation) === '.' ? process.cwd() : path.dirname(path.resolve(outLocation));
```
Comment (franvarney) [inline lib/controllers/project.js]: This is hard to read. Can we change to:

```
if (path.dirname(outLocation) === '.') dir = process.cwd();
else dir = path.dirname(path.resolve(outLocation));
```

(Or similar)

**Comment by franvarney** [comment]
:anchor:

File: lib/commands/project.js [hunk 3]
```diff
@@ -1042,6 +1041,48 @@ project.getMainFile = function(cb) {
...
+      userId: userConfig.data.userId
+    },
+    function (err, projects) {
+      if (err) {
+        err = error.handleApiError(err, 'GET_PROJECT_SOURCE', cb);
+        if (err.length > 0) return cb(err);
+      }
+
+      if (projects.length === 0) {
+        modulus.io.error('You currently have no projects. You can create one with "project create".');
```
Comment (harlanj) [inline lib/commands/project.js]: This is not a valid error for this request

**Comment by musgravejw** [comment]
:+1:

File: lib/commands/project.js (hunk 3 above)
Comment (tzmanics) [inline lib/commands/project.js]: This seems like a valid error message for `projects.length === 0`, am I missing something?

Comment (harlanj) [inline lib/commands/project.js]: Ah you're correct! I interpreted that incorrectly!

**Comment by tzmanics** [comment]
Tested this locally, worked :+1:

**Pull request merged**